### PR TITLE
feat(orc8r): reindexer singleton getBatches on new getJobs to reindex

### LINKED ISSUE #10135: Make reindexer a singleton service

### PR DESCRIPTION
**Pull request by reginawang3495**
## Summary

`reindexer_singleton` `getBatches` on new jobs to reindex
- previously, we only got batches on initialization (or `Run`) of the `reindexer_singleton` service
## Test Plan

Added a regression tests and ran all `orc8r` tests

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

File: orc8r/cloud/go/services/state/indexer/reindex/reindexer_singleton.go [hunk 1]
```diff
@@ -48,21 +48,33 @@ func (r *reindexerSingleton) Run(ctx context.Context) {
...
-	batches := r.getReindexBatches(ctx)
 	for {
 		if isCanceled(ctx) {
 			glog.Warning("State reindexing async job canceled")
 			return
 		}
 
-		r.reindexJobs(ctx, indexerID, batches, sendUpdate)
+		jobs, err := r.getJobs(indexerID)
+		if err == nil && jobs != nil {
```
Comment (hcgatewood) [inline orc8r/cloud/go/services/state/indexer/reindex/reindexer_singleton.go]: If there's an err, we want to log it

File: orc8r/cloud/go/services/state/indexer/reindex/reindexer_singleton.go [hunk 2]
```diff
@@ -48,21 +48,33 @@ func (r *reindexerSingleton) Run(ctx context.Context) {
 	// TODO(reginawang3495) will support non-nil sendUpdate from Run after removing queue impl
 	var sendUpdate func(string) = nil
 
-	batches := r.getReindexBatches(ctx)
 	for {
```
Comment (hcgatewood) [inline orc8r/cloud/go/services/state/indexer/reindex/reindexer_singleton.go:51]: For infinite loops, a good pattern is: 1. an outer func that creates the infinite loop, then, 2. a single func called within the loop. Makes validation and conceptualization easier

So in that case, it would be: let's make inside the for loop just be 3 things

1. is canceled
2. reindex jobs
3. log any err from reindex jobs

This way, reindexJobs will handle creating its own jobs, rather than taking it as an arg

File: orc8r/cloud/go/services/state/indexer/reindex/reindexer_singleton_test.go [hunk 3]
```diff
@@ -123,27 +124,47 @@ func TestSingletonRun(t *testing.T) {
 	fail3.AssertExpectations(t)
 	require.Equal(t, 2, reindexSuccessNum)
 	require.Equal(t, 5, reindexDoneNum)
+
+	// Test that a network/hardware pair that has been added after Run
+	// will have its states reindexed as well
```
Comment (hcgatewood) [inline orc8r/cloud/go/services/state/indexer/reindex/reindexer_singleton_test.go]: Add a newline under this comment, for clarity that it applies to the block rather than the line

File: orc8r/cloud/go/services/state/indexer/reindex/reindexer_singleton_test.go [hunk 4]
```diff
@@ -123,27 +124,47 @@ func TestSingletonRun(t *testing.T) {
...
+	addNetworks(t, reindexer)
+
+	idx5 := getIndexerNoIndex(id5, zero, indexer.Version(version5), true)
+	idx5.On("GetTypes").Return(allTypes).Once()
+	idx5.On("Index", mock.Anything, mock.Anything).Return(nil, nil).Times(4 * 3)
```
Comment (hcgatewood) [inline orc8r/cloud/go/services/state/indexer/reindex/reindexer_singleton_test.go]: This number's a little too magic, even for tests -- why 4*3 ?

File: orc8r/cloud/go/services/state/indexer/reindex/reindexer_singleton_test.go [hunk 5]
```diff
@@ -163,5 +184,39 @@ func initSingletonReindexTest(t *testing.T) reindex.Reindexer {
...
 	}
 
-	return reindexer
+	ctx, cancel := context.WithCancel(context.Background())
+	go reindexer.Run(ctx)
+
+	return cancel, reindexer
+}
+
+func addNetworks(t *testing.T, reindexer reindex.Reindexer) {
```
Comment (hcgatewood) [inline orc8r/cloud/go/services/state/indexer/reindex/reindexer_singleton_test.go]: "reportMoreState"

File: orc8r/cloud/go/services/state/indexer/reindex/reindexer_singleton_test.go [hunk 6]
```diff
@@ -163,5 +184,39 @@ func initSingletonReindexTest(t *testing.T) reindex.Reindexer {
...
+	ctx, _ := context.WithCancel(context.Background())
+	go reindexer.Run(ctx)
```
Comment (hcgatewood) [inline orc8r/cloud/go/services/state/indexer/reindex/reindexer_singleton_test.go]: Hoist this outside this func -- running this indexer is unrelated to reporting more state

**Review by hcgatewood — CHANGES_REQUESTED** [review]
Shape looks good! Couple small requests then I think we'll be ready to merge

File: orc8r/cloud/go/services/state/indexer/reindex/reindexer_singleton_test.go (hunk 6 above)
Comment (reginawang3495) [reply orc8r/cloud/go/services/state/indexer/reindex/reindexer_singleton_test.go]: It is related since we run `Run` before reporting more state

Comment (reginawang3495) [reply orc8r/cloud/go/services/state/indexer/reindex/reindexer_singleton_test.go]: Ahhh I see what you mean

File: orc8r/cloud/go/services/state/indexer/reindex/reindexer_singleton_test.go [hunk 7]
```diff
@@ -123,27 +124,53 @@ func TestSingletonRun(t *testing.T) {
 	fail3.AssertExpectations(t)
 	require.Equal(t, 2, reindexSuccessNum)
 	require.Equal(t, 5, reindexDoneNum)
+
+	// Test that a network/hardware pair that has been added after Run
+	// will have its states reindexed as well
+
+	ctx, _ := context.WithCancel(context.Background())
+	go reindexer.Run(ctx)
+
```
Comment (hcgatewood) [inline orc8r/cloud/go/services/state/indexer/reindex/reindexer_singleton_test.go]: Unless I'm misunderstanding: we shouldn't need to re-run the reindexer, because it's already running

File: orc8r/cloud/go/services/state/indexer/reindex/reindexer_singleton_test.go [hunk 8]
```diff
@@ -115,6 +113,9 @@ func TestSingletonRun(t *testing.T) {
 	recvCh(t, ch)
 	recvCh(t, ch)
 	recvCh(t, ch)
+
+	// Unregister indexers
+	register(t)
```
Comment (hcgatewood) [inline orc8r/cloud/go/services/state/indexer/reindex/reindexer_singleton_test.go]: Instead of doing this here, let's register at the beginning of your changes (below)

File: orc8r/cloud/go/services/state/indexer/reindex/reindexer_singleton_test.go [hunk 9]
```diff
@@ -163,5 +190,36 @@ func initSingletonReindexTest(t *testing.T) reindex.Reindexer {
...
+
+func reportMoreState(t *testing.T) {
+	configurator_test.RegisterNetwork(t, nid3, "Network 3 for reindex test")
+	configurator_test.RegisterGateway(t, nid3, hwid3, &models.GatewayDevice{HardwareID: hwid3})
+
+	ctxByNetwork := map[string]context.Context{
+		nid3: state_test.GetContextWithCertificate(t, hwid3),
+	}
+
+	// Report enough directory records to cause 3 batches per network (with the +1 gateway status per network)
```
Comment (hcgatewood) [inline orc8r/cloud/go/services/state/indexer/reindex/reindexer_singleton_test.go]: Move this comment to a func doc comment

**Review by hcgatewood — CHANGES_REQUESTED** [review]
Couple small things, otherwise lgtm

File: orc8r/cloud/go/services/state/indexer/reindex/reindexer_singleton_test.go (hunk 7 above)
Comment (reginawang3495) [reply orc8r/cloud/go/services/state/indexer/reindex/reindexer_singleton_test.go]: I cancelled it in L119 because of the purposefully failing indexers we were running

Comment (hcgatewood) [reply orc8r/cloud/go/services/state/indexer/reindex/reindexer_singleton_test.go]: Ah I see. Okay so we want to find a way to test this without canceling+restarting the reindexer -- to ensure that it gracefully handles newly-reported state without requiring a restart

We can either remove the `cancel()`, or move these "more state" checks before checking the failed indexers

Comment (reginawang3495) [reply orc8r/cloud/go/services/state/indexer/reindex/reindexer_singleton_test.go]: Ok sounds good, I'll try moving before the failing indexers

File: orc8r/cloud/go/services/state/indexer/reindex/reindexer_queue_test.go [hunk 10]
```diff
@@ -64,21 +64,27 @@ const (
 	nStatesToReindexPerCall    = 100 // copied from reindex.go
 	directoryRecordsPerNetwork = 2 * nStatesToReindexPerCall
 	nNetworks                  = 3
-	nBatches                   = 9
+	// 3 networks and 3 batches per network = 3 * 3 = 9
+	nBatches = 9
+	// 4 networks and 3 batches per network = 4 * 3 = 12
```
Comment (hcgatewood) [inline orc8r/cloud/go/services/state/indexer/reindex/reindexer_queue_test.go:69]: 👍

File: orc8r/cloud/go/services/state/indexer/reindex/reindexer_singleton.go [hunk 11]
```diff
@@ -48,21 +48,44 @@ func (r *reindexerSingleton) Run(ctx context.Context) {
...
+	if err != nil {
+		return wrap(err, ErrDefault, indexerID)
+	}
+	if jobs == nil {
+		return nil
+	}
+
 	batches := r.getReindexBatches(ctx)
-	return r.reindexJobs(ctx, indexerID, batches, sendUpdate)
+	return r.reindexJobs(ctx, jobs, batches, sendUpdate)
```
Comment (hcgatewood) [inline orc8r/cloud/go/services/state/indexer/reindex/reindexer_singleton.go:76]: This can probably just be "return find and reindex jobs". But is fine cause we're gonna remove this method anyway

**Review by hcgatewood — APPROVED** [review]
Looks great!
